### ISSUE TITLE
negtative coordinate

### ISSUE THREAD

**Issue by cyy21** [PeizeSun/SparseR-CNN#15]
the out_coord has negtive_value, which makes the loss l1 and giou very big, is it common? Did i miss some operations about out_coord, like sigmoid in detr?
i look into the code, the negtive value is from apply_deltas, 
pred_boxes[:, 0::4] = pred_ctr_x - 0.5 * pred_w  # x1
pred_boxes[:, 1::4] = pred_ctr_y - 0.5 * pred_h  # y1
results in negtive value

**Comment by PeizeSun** [comment]
Hi~ The out_coord may be negative value, but it won't cause l1 loss and giou loss very big. 
Do you change the code?

**Comment by cyy21** [comment]
@PeizeSun  thank you for reply, i'm a vegetable bird, and want to use sparse r-cnn train a custom dataset to publish a 中文核心 to graduate,the dataset has only one class. the raw code is under ubuntu, so i just copy the sparse r-cnn code to run, ididn't change the main code,just rewrite the fpn and roi with torchvision to get rid of detectron2(i have no machine to study cv, i can only use colab, my teacher know nothing about cv, just saying innovation everyday) ,i overfit a img with 5 gtbbox,1 class, i set adamw lr=0.00000015,i train for 20 epoch and the match indice and loss dict(has already mul weight) is
`[(array([ 9, 21, 31, 41, 91], dtype=int64), array([0, 4, 3, 2, 1]))]
{'loss_ce': tensor(1.1380, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([2.3593], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.8425], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 9, 17, 21, 31, 41], dtype=int64), array([0, 3, 4, 1, 2]))]
{'loss_ce': tensor(1.0639, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([2.3708], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.4607], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 9, 21, 31, 40, 41], dtype=int64), array([2, 4, 1, 0, 3]))]
{'loss_ce': tensor(1.0114, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([2.1171], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.5648], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 9, 21, 31, 40, 41], dtype=int64), array([2, 4, 1, 0, 3]))]
{'loss_ce': tensor(1.0618, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([1.9002], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.3918], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 9, 17, 21, 41, 91], dtype=int64), array([2, 0, 4, 3, 1]))]
{'loss_ce': tensor(1.2166, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([1.7000], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.2908], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 9, 17, 21, 31, 41], dtype=int64), array([2, 0, 4, 1, 3]))]
{'loss_ce': tensor(0.9589, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([1.9504], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.1862], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 9, 17, 21, 31, 41], dtype=int64), array([0, 1, 4, 3, 2]))]
{'loss_ce': tensor(0.8742, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([2.6168], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.6850], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 9, 17, 21, 31, 41], dtype=int64), array([0, 1, 2, 3, 4]))]
{'loss_ce': tensor(0.9568, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([1.9635], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.4190], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 9, 17, 19, 21, 31], dtype=int64), array([0, 1, 2, 4, 3]))]
{'loss_ce': tensor(0.9483, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([1.7790], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.6101], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 9, 17, 21, 41, 55], dtype=int64), array([0, 1, 4, 3, 2]))]
{'loss_ce': tensor(1.0138, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([1.7608], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.3643], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 5,  9, 17, 21, 41], dtype=int64), array([1, 0, 3, 4, 2]))]
{'loss_ce': tensor(0.8705, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([1.7089], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.2914], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 9, 17, 21, 31, 40], dtype=int64), array([2, 1, 4, 3, 0]))]
{'loss_ce': tensor(0.8886, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([2.1799], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.6958], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 9, 12, 17, 21, 41], dtype=int64), array([0, 2, 1, 4, 3]))]
{'loss_ce': tensor(0.9544, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([2.1234], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.6292], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([12, 17, 21, 41, 91], dtype=int64), array([2, 0, 4, 3, 1]))]
{'loss_ce': tensor(1.0589, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([2.4615], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.7692], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 5, 17, 21, 31, 41], dtype=int64), array([1, 0, 4, 3, 2]))]
{'loss_ce': tensor(0.9310, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([2.4576], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.5068], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 5, 17, 21, 31, 41], dtype=int64), array([1, 0, 4, 3, 2]))]
{'loss_ce': tensor(0.9330, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([1.9206], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.4372], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([17, 21, 31, 41, 91], dtype=int64), array([0, 4, 3, 2, 1]))]
{'loss_ce': tensor(1.0705, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([2.1755], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.5313], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 5, 17, 19, 21, 31], dtype=int64), array([1, 0, 2, 4, 3]))]
{'loss_ce': tensor(0.9544, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([1.7606], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.4117], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([17, 19, 21, 31, 91], dtype=int64), array([0, 2, 4, 3, 1]))]
{'loss_ce': tensor(1.1110, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([1.6532], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.4057], device='cuda:0', grad_fn=<MulBackward0>)}
[(array([ 5, 12, 17, 21, 31], dtype=int64), array([1, 2, 0, 4, 3]))]
{'loss_ce': tensor(1.0406, device='cuda:0', grad_fn=<MulBackward0>), 'loss_bbox': tensor([1.9994], device='cuda:0', grad_fn=<MulBackward0>), 'loss_giou': tensor([2.5561], device='cuda:0', grad_fn=<MulBackward0>)}

Process finished with exit code 0
`
i see the match is unstable, i think it is the reason why it cannot converge, i'm so vegitable, hope your suggestion, god pei

**Comment by PeizeSun** [comment]
Oh, your teacher looks bad bad~
Can you send your Wechat to my E-mail: sunpeize@foxmail.com, and we can discuss in Wechat.

I will close this issue.

**Comment by tyshiwo1** [comment]
I am a newbie in object detection area. I encounter the similer situation when I apply Sparse RCNN to other backbone. Can I send my Wechat and discuss? Thank you.

**Comment by cyy21** [comment]
@tyshiwo1 i think we can discuss with each other, my email is cyygwy@gmail.com
